cpp/include/cuspatial/experimental/detail/functors.cuh

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/cuda_utils.hpp>
+#include <cuspatial/experimental/geometry/segment.cuh>
+#include <cuspatial/traits.hpp>
+
+#include <thrust/binary_search.h>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @brief Given iterator a pair of offsets, return the number of elements between the offsets.
+ *
+ * Example:
+ * pair of offsets: (0, 3), (3, 5), (5, 8)
+ * number of elements between offsets: 3, 2, 3
+ *
+ * @tparam OffsetPairIterator Must be iterator type to thrust::pair of indices.
+ * @param p Iterator of thrust::pair of indices.
+ */
+struct offset_pair_to_count_functor {
+  template <typename OffsetPairIterator>
+  CUSPATIAL_HOST_DEVICE auto operator()(OffsetPairIterator p)
+  {
+    return thrust::get<1>(p) - thrust::get<0>(p);
+  }
+};
+
+/**
+ * @brief Convert counts of points to counts of segments in a linestring.
+ *
+ * A Multilinestring is composed of a series of Linestrings. Each Linestring is composed of a
+ * segments. The number of segments in a multilinestring is the number of points in the
+ * multilinestring subtracting the number of linestrings.
+ *
+ * Caveats: This has a strong assumption that the Multilinestring does not contain empty linestring.
+ * While each non-empty linestring in the multilinestring can cause 1 invalid segment, an empty
+ * multilinestring not introduce and invalid segments since it does not contain any points.
```

Comment:
```suggestion
 * Caveats: This has a strong assumption that the Multilinestring does not contain empty linestrings.
 * While each non-empty linestring in the multilinestring represents 1 extra segment, an empty
 * multilinestring does not introduce any extra segments (since it does not contain any points).
```